runtime/wasm/src/module/mod.rs

```diff
@@ -652,30 +657,56 @@ impl<C: Blockchain> AscHeap for WasmInstanceContext<C> {
         Ok(ptr as u32)
     }
 
-    fn get(
-        &self,
+    fn read_u32(&self, offset: u32, gas: &GasCounter) -> Result<u32, DeterministicHostError> {
+        gas.consume_host_fn(Gas::new(GAS_COST_LOAD as u64 * 4))?;
+        let mut bytes = [0; 4];
+        self.memory.read(offset as usize, &mut bytes).map_err(|_| {
+            DeterministicHostError::from(anyhow!(
+                "Heap access out of bounds. Offset: {} Size: {}",
+                offset,
+                4
+            ))
+        })?;
+        Ok(u32::from_le_bytes(bytes))
+    }
+
+    fn init<'s, 'a>(
+        &'s self,
         offset: u32,
-        size: u32,
+        buffer: &'a mut [MaybeUninit<u8>],
         gas: &GasCounter,
-    ) -> Result<Vec<u8>, DeterministicHostError> {
+    ) -> Result<&'a mut [u8], DeterministicHostError> {
         // The cost of reading wasm memory from the host is the same as of reading from wasm using
         // load instructions.
-        gas.consume_host_fn(Gas::new(GAS_COST_LOAD as u64 * size as u64))?;
+        gas.consume_host_fn(Gas::new(GAS_COST_LOAD as u64 * (buffer.len() as u64)))?;
 
         let offset = offset as usize;
-        let size = size as usize;
 
-        let mut data = vec![0; size];
-
-        self.memory.read(offset, &mut data).map_err(|_| {
-            DeterministicHostError::from(anyhow!(
-                "Heap access out of bounds. Offset: {} Size: {}",
-                offset,
-                size
-            ))
-        })?;
-
-        Ok(data)
+        unsafe {
+            // Safety: This was copy-pasted from Memory::read, and we ensure
+            // nothing else is writing this memory because we don't call into
+            // WASM here.
+            let src = self
+                .memory
+                .data_unchecked()
+                .get(offset..)
+                .and_then(|s| s.get(..buffer.len()))
+                .ok_or(DeterministicHostError::from(anyhow!(
+                    "Heap access out of bounds. Offset: {} Size: {}",
+                    offset,
+                    buffer.len()
+                )))?;
+
+            // See also 6d2f040c-9361-4cf9-8b3c-426f87cc0b21
+            // When MaybeUninit::write_slice is stabilized, this won't
+            // require any unsafe code and will be a one-liner
+            let uninit_src: &[MaybeUninit<u8>] = std::mem::transmute(src);
+            buffer.copy_from_slice(uninit_src);
+            // This is copied from slice_assume_init_mut, which is
+            // currently an unstable API
+            let buffer = &mut *(buffer as *mut [MaybeUninit<u8>] as *mut [u8]);
```

Comment:
This is copy pasted in two places, could it be extracted to a safe utility function somewhere in `graph`?

Reply:
Done